### ISSUE TITLE
SystemJS 0.20 support

### ISSUE THREAD

**Issue by guybedford** [frankwallis/plugin-typescript#185]
SystemJS 0.20 is coming out tomorrow, and seems to break this plugin at the moment.

The reason for this is that `import * as ts from 'typescript'` used in the plugin itself, which is no longer being supported for loading from legacy module formats. This is to align with compatibility with NodeJS (see https://github.com/systemjs/systemjs/releases/tag/0.20.0-rc.2).

A temporary fix is to do:

typescript-shim.js
```javascript
var ts = require('actual-typescript');
ts.__esModule = true;
module.exports = ts;
```

with the configuration:

```javascript
		SystemJS.config({
			packages: {
				"ts": {
					"main": "plugin.js"
				},
				"actual-typescript": {
					"main": "lib/typescript.js",
					"meta": {
						"lib/typescript.js": {
							"exports": "ts"
						}
					}
				}
			},
			map: {
				"ts": "../node_modules/plugin-typescript/lib",
				"actual-typescript": "../node_modules/typescript",
				"typescript": "./typescript-shim.js"
			},
			transpiler: 'ts'
		});
```

But the better fix is of course to update the built output to use the `import ts from 'typescript'` form. I tried changing this but got a build error - perhaps the synthetic default imports option needs to be enabled?

Just let me know if I can help further here at all.

**Comment by frankwallis** [comment]
@guybedford I have a solution for this (the one you describe above) which I will release, but I am seeing another issue at the moment which is that ```SystemJS.typescriptOptions``` is no longer getting populated from the config.

I can thread it though by setting it in metadata and accessing ```load.metadata.typescriptOptions``` for each file, but this will be a big breaking change for consumers. Looking at https://github.com/systemjs/plugin-babel/blob/master/plugin-babel.js#L52 it sounds like it should be available on ```loader``` but it is not coming through? Please advise, thanks.

**Comment by guybedford** [comment]
@frankwallis it is still possible to set `System.typescriptOptions` directly on the System object, but just not via the `System.config` call as this is now locked to internal state options only. So yes using the meta configuration or explicit setting is probably best. This was necessary so that we can start validating configuration options for SystemJS in the next release.

**Comment by guybedford** [comment]
And thanks so much for looking into these changes, it is really appreciated, just let me know if you have any other questions about the release.

**Comment by filipesilva** [comment]
@frankwallis I've a vested interest in making sure this works since https://github.com/angular/quickstart and https://github.com/angular/angular.io heavily depend on it for examples. Let me know if there's anything I can do to help.

**Comment by filipesilva** [comment]
I've tried updating one of our plunkers to use `systemjs@0.20.0` and `plugin-typescript@6.0.1` (from git), and seem to be stuck with the following error:

https://plnkr.co/edit/fR7gMq8SsW8GyPIx10DK
```
Error: core_1.Component is not a function
  Evaluating https://run.plnkr.co/4QiG4lZEBjkUZafX/app/app.component.ts
  Evaluating https://run.plnkr.co/4QiG4lZEBjkUZafX/app/app.module.ts
  Evaluating https://run.plnkr.co/4QiG4lZEBjkUZafX/app/main.ts
  Loading app
    at eval (app.component.ts:3)
    at Object.execute (app.component.ts:7)
    at declarativeExecute (register-loader.js:649)
```

Where `app.component.ts`:
```
import { Component } from '@angular/core';

@Component({
  selector: 'my-app',
  template: `<h1>Hello {{name}}</h1>`
})
export class AppComponent { name = 'Angular'; }
```

I made sure the plunker worked correctly with no warnings using the previous latest versions (`systemjs@0.19.43` and `plugin-typescript@5.3.1`). Well there this single warning but it is intended, as we need `module.id`:
```
TypeScript transpiling to CommonJS, consider setting module: "system" in typescriptOptions to transpile directly to System.register format
```

Is this an error in our configuration, or is `plugin-typescript@6.0.1` currently incomplete?

**Comment by guybedford** [comment]
It sounds like this is a result of removing named exports for legacy
formats.

Typescript doesn't add the __esModule flag to modules I believe, so that
would be the reason for the named export no longer working.

The fix would be either to compile to the System format, ensure the
__esModule flag is exported before compilation to make sure it is there, or
to use the default import only.

Let me know if that makes sense.

**Comment by filipesilva** [comment]
Using the default import only would be a big problem for us since that implies a stark difference between local and plunker setups.

We use `module.id` both in plunkers and local setups to ensure a fairly homogeneous setup, and I think compiling directly to the System format does not preserve it. Correct me if I'm wrong though.

So out of the three options, it sounds like having __esModule flag exported is the least disruptive option for our setups. Is this something that can be done on the config?

**Comment by guybedford** [comment]
Yeah it's the pain of accepting CommonJS modules are only seen as a default export from ES modules.

The best option then may just be to create a wrapper module, just like I created one for TypeScript above in https://github.com/frankwallis/plugin-typescript/issues/185#issue-202481354. The wrapper module does need to be written as code though, so makes the plunkr a little more complex unfortunately.

**Comment by filipesilva** [comment]
I tried that approach but it seems to give me the exact same error. I also tried with the previous version of `plugin-typescript` and it seems the same. 

Am I missing something?

https://plnkr.co/edit/nXgx15mhMNYGPkU7RUa5?p=preview
```
// typescript-shim.js
var ts = require('actual-typescript');
ts.__esModule = true;
module.exports = ts;
```
```
// config
    meta: {
      'actual-typescript': {
        "exports": "ts"
      }
    },
// ... 
    map : {
    //...
      'ts': 'https://cdn.rawgit.com/frankwallis/plugin-typescript/f6a24f13/lib/plugin.js',
      'actual-typescript': 'npm:typescript@2.0.3/lib/typescript.js',
      'typescript': './typescript-shim.js',
```

**Comment by guybedford** [comment]
@filipesilva so yes it's exactly the same technique, but copying it for loading angular/core itself so that the named exports can work. Just looking at it now, you may well find that the same applies to all the other builds which may well be unwieldy...

**Comment by filipesilva** [comment]
Oh my... that is problematic. I see now. So any packages via `plugin-typescript`, while configured with `"module": "commonjs"`, would need a shim.

That makes me think I need to take another approach as we couldn't expect plunker users to do that for other packages they might want to add.

That made me try `"module": "system"` but that gave me the same error still:
https://plnkr.co/edit/kzI0Up2Y0Bnsim4xg7ge?p=preview

Perhaps there is something still missing in `plugin-typescript`?

**Comment by guybedford** [comment]
@filipesilva I don't think this issue is specifically about plugin-typescript, it is just about how to load a CommonJS module from an ES module in SystemJS. 

The issue being that because Angular has been compiled into UMD, it can only be imported by ES modules as a default import, so it's about how the dependency is interpreted. Basically we just need all the UMD builds to have an `__esModule: true` flag set on them to tell SystemJS not to treat them with the correct interop, but as "fake ES modules".

We could potentially add an option to SystemJS to inform when modules should be treated as if they have an esModule flag, but via meta configuration. This would make the configuration here:

```javascript
System.config({
  meta: {
    '@angular/core': { esModule: true },
    '@angular/common': { esModule: true },
    ... etc
  }
});
```

I can post out a patch with this fix soon if you think that would be best, but perhaps just confirm there really is no way the `__esModule` flag could be exposed on these builds?

**Comment by guybedford** [comment]
Something like `esModuleCompat` is probably a better name.

**Comment by filipesilva** [comment]
@guybedford you're right. I'm still getting to terms with the full implications of this change. This would affect both TS and ES6 transpilation then.

I'm going to pass on the proposed UMD bundle change to the core team and direct them here.

Your writeup in http://guybedford.com/systemjs-alignment#node-modules-compatibility is helping a lot, thank you.

**Comment by guybedford** [comment]
@filipesilva ok glad it is making sense. I've put together a branch along the lines I suggested at https://github.com/systemjs/systemjs/tree/angular-esmodule, let me know if you think this route is worth taking as well.

**Comment by guybedford** [comment]
@frankwallis thanks again for looking into this... also if you feel the current situation is overly constrained do let me know your thoughts too.

**Comment by filipesilva** [comment]
@guybedford thanks for making that branch!

I tried testing with it but I seem to run into the same error again:
(note: I had to fork and build it myself, your changes weren't in `dist/`)

https://plnkr.co/edit/bLS0j6wQ7qYeE2jWBYn6?p=preview

`Error: core_1.Component is not a function`

```
    meta: {
      '@angular/core': { esModule: true },
      '@angular/common': { esModule: true },
      '@angular/compiler': { esModule: true },
      '@angular/platform-browser': { esModule: true },
      '@angular/platform-browser-dynamic': { esModule: true },
      '@angular/http': { esModule: true },
      '@angular/router': { esModule: true },
      '@angular/forms': { esModule: true },
      '@angular/upgrade': { esModule: true },
      '@angular/upgrade/static': { esModule: true },
      //...
```

I tried debugging a bit with a local env and the code from your test:
```
return System.import('@angular/core').then(function () {
      var m = System.registry.get(System.resolveSync('@angular/core'));
      console.log(m);
    });
```

I found that the `@angular/core` call didn't seem to go through `setMetaEsModule()`... but `rxjs` did. Am I missing something?

**Comment by guybedford** [comment]
@filipesilva are the Angular builds perhaps running as AMD modules? I left out the AMD format support for the `esModule` feature in that experimental branch, as it's a little more work to get it going for that case due to the indirection of the AMD define.

**Comment by filipesilva** [comment]
@guybedford something else occurred to me regarding the `esModule` flag: although UMD libraries can include it (Rollup has done it for a while), locally transpiled files cannot. 

So in my specific case, the Typescript app being compiled to commonjs on the browser would always need an option in SystemJS to force `esModule`.

Is this reasoning correct?

**Comment by filipesilva** [comment]
I'm not sure whether or not they get detected as AMD to be honest... I know they were build with rollup@0.26.x and look like this:
https://unpkg.com/@angular/core@2.4.5/bundles/core.umd.js

It seems like they'd be able to be loaded as either CommonJS or AMD, but I don't know what ends up happening.

**Comment by filipesilva** [comment]
@frankwallis in 6.0.1 it seems that `typescriptOptions` options are ignored. Setting:

```
    typescriptOptions: {
      tsconfig: './nothing-here.json'  
    },
```

Does not result in an error.

**Comment by guybedford** [comment]
@filipesilva try setting `format: 'cjs'` as well as the `esModule: true` flag in the meta for the UMD builds. The combination may work as a temporary fix.

**Comment by guybedford** [comment]
@filipesilva your other point sounds like it is related to https://github.com/frankwallis/plugin-typescript/issues/185#issuecomment-274902258.

**Comment by filipesilva** [comment]
`format: 'cjs'` did the trick! App loads correctly now:
```
      '@angular/core': { esModule: true, format: 'cjs' },
      '@angular/common': { esModule: true, format: 'cjs' },
      '@angular/compiler': { esModule: true, format: 'cjs' },
      '@angular/platform-browser': { esModule: true, format: 'cjs' },
      '@angular/platform-browser-dynamic': { esModule: true, format: 'cjs' },
      '@angular/http': { esModule: true, format: 'cjs' },
      '@angular/router': { esModule: true, format: 'cjs' },
      '@angular/forms': { esModule: true, format: 'cjs' },
      '@angular/upgrade': { esModule: true, format: 'cjs' },
      '@angular/upgrade/static': { esModule: true, format: 'cjs' },
```

I went back to the comment you mentioned and was able to set the options via `SystemJS.typescriptOptions`:
```
  SystemJS.typescriptOptions = {
    // Copy of compiler options in standard tsconfig.json
    "target": "es5",
    "module": "commonjs",
    "moduleResolution": "node",
    "sourceMap": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "noImplicitAny": true,
    "suppressImplicitAnyIndexErrors": true
  };
```
I could verify `console.log(module.id)` works.

So at the moment I can say that the `esModule: true` flag supports our usecase. 

I'm going back with these findings to see if it's possible to include the `__esModule` flag in the UMDs since it seems that's really the only thing needed here, as far as I can tell.

**Comment by filipesilva** [comment]
@guybedford I just talked with @wardbell and he is of the opinion that it would better to have both the meta flag on the SystemJS and for Angular bundles to also include it.

One of the big motivators for this approach is that existing, published older versions of Angular do not have this flag and thus would not be usable.

What do you think?

**Comment by guybedford** [comment]
@filipesilva sure, let me work on getting this flag into a patch release.

**Comment by guybedford** [comment]
@filipesilva I've released a SystemJS 0.20.2 with the `esModule` flag, including support for it for AMD modules, so the `format: 'cjs'` meta will no longer be needed. Hopefully that gets you there now!?

**Comment by filipesilva** [comment]
@guybedford yes, very much so! I used the new patch version, removed the `format: 'cjs'`, and can verify it works as expected:
https://plnkr.co/edit/uyIdXSoE90NO57QY2hRm?p=preview

I'll working on getting the `__esModule` flag into https://github.com/angular/angular. together with tests.

Thank you for being so helpful in getting this sorted @guybedford!

**Comment by guybedford** [comment]
@filipesilva that's great news, thanks for working on upgrading there so quickly too.

**Comment by frankwallis** [comment]
Also worth noting that there are plans to add the __esModule flag to TypeScript transpiled output, (see https://github.com/Microsoft/TypeScript/issues/13709).

I am closing this now and will update the examples and docs to set ```typescriptOptions``` directly. @guybedford - would it be possible to add a warning if the user tries to set ```typescriptOptions``` via System.config as currently it just gets ignored and that will be a source of issues as people upgrade?

**Comment by guybedford** [comment]
@frankwallis I've undeprecated the ability to use `SystemJS.config({ typescriptOptions: {...} })` in SystemJS 0.20.3... hopefully that helps!

**Comment by frankwallis** [comment]
@guybedford thank you, that is a massive help

**Comment by filipesilva** [comment]
@frankwallis are there plans to release `6.0.1` outside of github at the moment, or will a future release only be on npm?

**Comment by frankwallis** [comment]
@filipesilva I have released 6.0.4 on github and npm

**Comment by filipesilva** [comment]
@frankwallis awesome, I'll start using it right away!
